applications/FluidDynamicsApplication/custom_elements/data_containers/derivatives.h

```diff
@@ -0,0 +1,172 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_DERIVATIVES_H)
+#define KRATOS_DERIVATIVES_H
+
+// System includes
+#include <tuple>
+#include <utility>
+
+// External includes
+
+// Project includes
+#include "includes/ublas_interface.h"
+
+// Application includes
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+template<unsigned int TNumNodes, unsigned int TBlockSize>
+class ZeroDerivatives
+{
+public:
+    ///@name Local type definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    constexpr static IndexType NumNodes = TNumNodes;
+
+    constexpr static IndexType BlockSize = TBlockSize;
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    template<IndexType TSize, class... TArgs>
+    void inline CalculateGaussPointResidualsDerivativeContributions(BoundedVector<double, TSize>& rResidualDerivative, TArgs&... rArgs) const
+    {
+        rResidualDerivative.clear();
+    }
+
+    ///@}
+};
+
+template<class TSubAssemblyType, unsigned int TElementDataHolderIndex, unsigned int TRowStartingIndex, unsigned int TColumnStartingIndex = 0>
+class SubAssembly : public TSubAssemblyType
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using SubAssemblyType = TSubAssemblyType;
+
+    static constexpr unsigned int ElementDataIndex = TElementDataHolderIndex;
+
+    static constexpr unsigned int RowStartingIndex = TRowStartingIndex;
+
+    static constexpr unsigned int ColumnStartingIndex = TColumnStartingIndex;
+
+    static constexpr unsigned int NumNodes = SubAssemblyType::NumNodes;
+
+    static constexpr unsigned int BlockSize = SubAssemblyType::BlockSize;
+
+    static constexpr unsigned int ResidualSize = NumNodes * BlockSize;
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    inline BoundedVector<double, ResidualSize>& GetSubVector()
+    {
+        return mSubVector;
+    }
+
+    template<class TCombinedElementDataContainer>
+    inline typename std::tuple_element<ElementDataIndex, TCombinedElementDataContainer>::type& GetElementDataContainer(TCombinedElementDataContainer& rCombinedDataContainer) const
+    {
+        static_assert(
```

Comment:
👍 Cool. I've the feeling that there should be several error macros that could be substituted by this.